ScoutingManager.cpp

```diff
@@ -0,0 +1,246 @@
+				// For each possible base, get the distance from enemy and add to total distance
+				for (auto pos : inv.expo_positions_complete_) {
+					int base_distance = inv.getRadialDistanceOutFromEnemy(Position(pos));
+					total_distance += base_distance;
+					scout_expo_distances_.push_back(base_distance);
+				}
+
+				int mean_base_distance = total_distance / size(scout_expo_distances_);
+				
+				// For each base distance, if that distance is smaller than the avg then that base is a more likely enemy base
+				for (auto dis : scout_expo_distances_) {
+					if (dis < mean_base_distance) {
+						// Reassign the base distances to actual expo positions
+						for (auto pos : inv.expo_positions_complete_) {
+							if (dis == inv.getRadialDistanceOutFromEnemy(Position(pos))) {
+								scout_expo_positions_.push_back(Position(pos));
+								break; //If we found the match, we are finished
+							}
+						}
+					}
+				}
+			}
+
+			scout_spot = scout_expo_positions_[0];
+			scout_expo_positions_.erase(scout_expo_positions_.begin());  // Scout goes to unique location each time
+			return scout_spot;
+		}
+	}
+	return scout_spot;
+}
+
+bool ScoutingManager::needScout(const Unit &unit, const int &t_game) const {
+// When do we want to scout
+	UnitType u_type = unit->getType();
+
+	// Make sure we have a zergling
+	if (u_type == UnitTypes::Zerg_Zergling) {
+		// Always have an expo scout
+		if (!exists_expo_zergling_scout_)
+			return true;
+		// If a suicidal zergling scout doesn't exists and it's been 30s since death
+		if (!exists_zergling_scout_ && last_zergling_scout_sent_ < t_game - Broodwar->getLatencyFrames() - 30 * 24)
+			return true;
+	}
+
+	// Make sure we have an overlord
+	if (u_type == UnitTypes::Zerg_Overlord) {
+		// Initial overlord scout
+		if (t_game < 3)
+			return true;
+		// If an overlord scout doesn't exist and it's been 60s since death
+		if (!exists_overlord_scout_ && last_overlord_scout_sent_ < t_game - Broodwar->getLatencyFrames() - 60 * 24)
+			return true;
+	}
+
+	// Don't need a scout
+	return false;
+}
+
+void ScoutingManager::updateScouts() {
+// Check if scouts have died
+
+	//if we thought we had a suicide zergling scout but now we don't
+	if (zergling_scout_ != nullptr) {
+		if (!zergling_scout_->exists()) {
+			zergling_scout_ = 0;
+			exists_zergling_scout_ = false;
+			last_zergling_scout_sent_ = Broodwar->getFrameCount();
+			Broodwar->sendText("Zergling scout died");
+		}
+	}
+
+	//if we thought we had an expo zergling scout but now we don't
+	if (expo_zergling_scout_ != nullptr) {
+		if (!expo_zergling_scout_->exists()) {
+			expo_zergling_scout_ = 0;
+			exists_expo_zergling_scout_ = false;
+			Broodwar->sendText("Expo ling scout died");
+		}
+	}
+
+	//if we thought we had an overlord scout but now we don't
+	if (overlord_scout_ != nullptr) {
+		if (!overlord_scout_->exists()) {  //if we thought we had a scout but now we don't
+			overlord_scout_ = 0;
```

Comment:
to be very crisp, you might want to set these to nullptr instead of 0. They are typed as units below and so setting them to zero is arguably inappropriate.